File: platform-auth-server/src/main/java/com/webank/wecube/platform/auth/server/service/AuthorityManagementService.java
package com.webank.wecube.platform.auth.server.service;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.webank.wecube.platform.auth.server.common.AuthServerException;
import com.webank.wecube.platform.auth.server.dto.SimpleAuthorityDto;
import com.webank.wecube.platform.auth.server.entity.SysAuthorityEntity;
import com.webank.wecube.platform.auth.server.http.AuthenticationContextHolder;
import com.webank.wecube.platform.auth.server.repository.AuthorityRepository;

@Service("authorityManagementService")
public class AuthorityManagementService {

    private static final Logger log = LoggerFactory.getLogger(AuthorityManagementService.class);

    @Autowired
    private AuthorityRepository authorityRepository;

    public SimpleAuthorityDto registerLocalAuthority(SimpleAuthorityDto authDto) {
        if (StringUtils.isBlank(authDto.getCode())) {
            throw new AuthServerException("3000", "Authority code to register cannot be blank.");
        }

        SysAuthorityEntity authority = authorityRepository.findNotDeletedOneByCode(authDto.getCode());

        if (authority != null) {
            log.debug("authority {} to register already exists.", authDto.getCode());
            String msg = String.format(
                    "Authority registering failed,because authority code {%s} already exist.", authDto.getCode());
            throw new AuthServerException("3001", msg, authDto.getCode());
        }

        authority = new SysAuthorityEntity();
        authority.setActive(true);
        authority.setCode(authDto.getCode());
        authority.setCreatedBy(AuthenticationContextHolder.getCurrentUsername());
        authority.setDeleted(false);
        authority.setDescription(authDto.getDescription());
        authority.setDisplayName(authDto.getDisplayName());
        authority.setScope(
                StringUtils.isBlank(authDto.getScope()) ? SysAuthorityEntity.SCOPE_GLOBAL : authDto.getScope());

        SysAuthorityEntity saveAuthority = authorityRepository.saveAndFlush(authority);

        return convertToSimpleAuthorityDto(saveAuthority);
    }
    
    private SimpleAuthorityDto convertToSimpleAuthorityDto(SysAuthorityEntity authority){
        SimpleAuthorityDto dto = new SimpleAuthorityDto();
        dto.setActive(authority.isActive());
        dto.setCode(authority.getCode());
        dto.setDescription(authority.getDescription());
        dto.setDisplayName(authority.getDisplayName());
        dto.setId(authority.getId());
        dto.setScope(authority.getScope());
        
        return dto;
    }

    public List<SimpleAuthorityDto> retrieveAllLocalAuthorites() {
        List<SimpleAuthorityDto> result = new ArrayList<>();
        List<SysAuthorityEntity> authorities = authorityRepository.findAllNotDeletedAuthorities();
        if(authorities == null || authorities.isEmpty()){
            return result;
        }
        
        for(SysAuthorityEntity authority : authorities){
            SimpleAuthorityDto dto = convertToSimpleAuthorityDto(authority);
            result.add(dto);
        }
        return result;
    }
}

File: platform-auth-server/service/authority_management_service.go
package service

import (
	"fmt"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/constant"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/exterror"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/log"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/utils"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/model"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/service/db"
	"time"
)

var AuthorityManagementServiceInstance AuthorityManagementService

type AuthorityManagementService struct {
}

func (a AuthorityManagementService) RegisterLocalAuthority(authDto model.SimpleAuthorityDto, curUser string) (*model.SimpleAuthorityDto, error) {
	if len(authDto.Code) == 0 {
		return nil, exterror.Catch(exterror.New().AuthServer3000Error, nil)
	}

	authority, err := db.AuthorityRepositoryInstance.FindNotDeletedOneByCode(authDto.Code)
	if err != nil {
		log.Logger.Warn("can not find not deleted auth by code", log.String("code", authDto.Code), log.Error(err))
		return nil, err
	}

	if authority != nil {
		log.Logger.Debug(fmt.Sprintf("authority %v to register already exists.", authDto.Code))
		/*		msg := fmt.Sprintf(
				"Authority registering failed,because authority code {%s} already exist.", authDto.Code);
		*/return nil, exterror.Catch(exterror.New().AuthServer3001Error.WithParam(authDto.Code), nil)
	}
	scope := authDto.Scope
	if len(scope) == 0 {
		scope = constant.SCOPE_GLOBAL
	}

	authority = &model.SysAuthorityEntity{
		Id:          utils.Uuid(),
		Active:      true,
		Code:        authDto.Code,
		CreatedBy:   curUser,
		Deleted:     false,
		Description: authDto.Description,
		DisplayName: authDto.DisplayName,
		Scope:       scope,
		CreatedTime: time.Now(),
	}
	db.Engine.Insert(authority)

	return convertToSimpleAuthorityDto(authority), nil

}

func convertToSimpleAuthorityDto(authority *model.SysAuthorityEntity) *model.SimpleAuthorityDto {
	return &model.SimpleAuthorityDto{
		Active:      authority.Active,
		Code:        authority.Code,
		Description: authority.Description,
		DisplayName: authority.DisplayName,
		ID:          authority.Id,
		Scope:       authority.Scope,
	}
}

func (a AuthorityManagementService) RetrieveAllLocalAuthorities() ([]*model.SimpleAuthorityDto, error) {
	result := make([]*model.SimpleAuthorityDto, 0)
	authorities, err := db.AuthorityRepositoryInstance.FindAllNotDeletedAuthorities()
	if err != nil {
		log.Logger.Error("failed to find all not deleted auth", log.Error(err))
		return nil, err
	}
	if len(authorities) == 0 {
		return result, nil
	}

	for _, authority := range authorities {
		dto := convertToSimpleAuthorityDto(authority)
		result = append(result, dto)
	}
	return result, nil
}
